### PR TITLE
Pushdown DATE, TIMESTAMP, and TIMESTAMP_TZ_MILLIS predicates in MongoDB

### PR DESCRIPTION
**Pull request by harlequin**
Resolves: #5658
Superseeds: #5682, because it seems that the developer is not active in the moment.

This PR takes all changes as in #5682 and implement additional the proposed changes from discussion.

### CODE REVIEW

**Review by ebyhr — COMMENTED** [review]
Please squash commits into one. I would recommend preserving commit author when picking up other contributor's commit.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 1]
```diff
@@ -508,6 +512,18 @@ private static Optional<Object> translateValue(Object trinoNativeValue, Type typ
             return Optional.of(trinoNativeValue);
         }
 
+        if (type == DATE) {
+            return Optional.of(new Date(TimeUnit.DAYS.toMillis((Long) trinoNativeValue)));
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:516]: I would recommend adding a preparatory commit to use `LocalDate` for DATE type in other places.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 2]
```diff
@@ -508,6 +512,18 @@ private static Optional<Object> translateValue(Object trinoNativeValue, Type typ
...
+        }
+
+        if (type == TIMESTAMP_MILLIS) {
+            return Optional.of(new Date(TimeUnit.MILLISECONDS.convert((Long) trinoNativeValue, TimeUnit.MICROSECONDS)));
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:520]: Use `LocalDateTime`

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 3]
```diff
@@ -508,6 +512,18 @@ private static Optional<Object> translateValue(Object trinoNativeValue, Type typ
...
+        }
+
+        if (type == TIMESTAMP_TZ_MILLIS) {
+            return Optional.of(new Date(unpackMillisUtc((long) trinoNativeValue)));
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:524]: Use `LocalDateTime`

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/BaseMongoConnectorTest.java [hunk 4]
```diff
@@ -589,6 +590,41 @@ protected void verifyTableNameLengthFailurePermissible(Throwable e)
...
+
+        // bigint equality
+        assertThat(query("SELECT regionkey, nationkey, name FROM nation WHERE nationkey = 19"))
+                .matches("VALUES (BIGINT '3', BIGINT '19', CAST('ROMANIA' AS varchar(25)))")
+                .isNotFullyPushedDown(FilterNode.class);
+
+        // bigint range, with decimal to bigint simplification
+        assertThat(query("SELECT regionkey, nationkey, name FROM nation WHERE nationkey BETWEEN 18.5 AND 19.5"))
+                .matches("VALUES (BIGINT '3', BIGINT '19', CAST('ROMANIA' AS varchar(25)))")
+                .isNotFullyPushedDown(FilterNode.class);
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/BaseMongoConnectorTest.java:620]: These tests unrelates to date, timestamp, timestamp_tz_millis change. Please separate a commit.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoTypeMapping.java [hunk 5]
```diff
@@ -251,6 +251,9 @@ public void testTime()
                 .addRoundTrip("time", "TIME '23:59:59.9'", createTimeType(3), "TIME '23:59:59.900'")
                 .addRoundTrip("time", "TIME '23:59:59.99'", createTimeType(3), "TIME '23:59:59.990'")
                 .addRoundTrip("time", "TIME '23:59:59.999'", createTimeType(3), "TIME '23:59:59.999'")
+                .addRoundTrip("time", "TIME '23:59:59.9999'", createTimeType(4), "TIME '23:59:59.9999'")
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoTypeMapping.java:254]: Add tests for timestamp and timestamp with time zone types.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoTypeMapping.java [hunk 6]
```diff
@@ -251,6 +251,9 @@ public void testTime()
...
+                .addRoundTrip("time", "TIME '23:59:59.99999'", createTimeType(5), "TIME '23:59:59.99999'")
+                .addRoundTrip("time", "TIME '23:59:59.999999'", createTimeType(6), "TIME '23:59:59.999999'")
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoTypeMapping.java:256]: Please check CI failures.

**Comment by yujiosaka** [comment]
hope to see some progress 🙏 
i also tried a bit but it was a bit difficult for me to proceed due to the lack of the domain knowledge

**Comment by ebyhr** [comment]
Superseded by #14413

**Pull request closed without merging**